pkg/addonmanager/controllers/agentdeploy/hosted_sync.go

```diff
@@ -80,37 +80,36 @@ func (s *hostedSyncer) sync(ctx context.Context,
 		Message: fmt.Sprintf("hosting cluster %s is a managed cluster of the hub", hostingClusterName),
 	})
 
-	if !hostingCluster.DeletionTimestamp.IsZero() {
-		if err = s.cleanupDeployWork(ctx, addon); err != nil {
-			return addon, err
+	// Don't skip syncing if the addon is deleting and there is a predelete hook, since the deployment manifests may
+	// need to be updated during the uninstall.
+	if !addonHasFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer) {
+		if !hostingCluster.DeletionTimestamp.IsZero() {
+			if err = s.cleanupDeployWork(ctx, addon); err != nil {
+				return addon, err
+			}
+			addonRemoveFinalizer(addon, addonapiv1alpha1.AddonHostingManifestFinalizer)
+			return addon, nil
 		}
-		addonRemoveFinalizer(addon, addonapiv1alpha1.AddonHostingManifestFinalizer)
-		return addon, nil
-	}
 
-	if !addon.DeletionTimestamp.IsZero() {
-		// clean up the deploy work until the hook work is completed
-		if addonHasFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer) {
+		if !addon.DeletionTimestamp.IsZero() {
+			if err = s.cleanupDeployWork(ctx, addon); err != nil {
+				return addon, err
+			}
+			addonRemoveFinalizer(addon, addonapiv1alpha1.AddonHostingManifestFinalizer)
 			return addon, nil
 		}
 
-		if err = s.cleanupDeployWork(ctx, addon); err != nil {
-			return addon, err
+		// waiting for the addon to be deleted when cluster is deleting.
+		// TODO: consider to delete addon in this scenario.
+		if !cluster.DeletionTimestamp.IsZero() && !addonHasFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer) {
```

Comment:
```suggestion
		if !cluster.DeletionTimestamp.IsZero() {
```

Reply:
@zhujian7 I wasn't sure about this because I could see the cluster being deleted, which causes the addon to be uninstalled, and you may still want to deploy new changes to the deployment.

In the policy use-case, we want to set an annotation on the `Deployment` object to indicate it is being uninstalled and be able to deploy new image updates if the uninstall is stuck and requires a bug fix. I think we would still need to wait for the `AddonHostingPreDeleteHookFinalizer` finalizer to be removed for this to still work when the whole cluster is being uninstalled scenario.

Reply:
yes, thanks @mprahl for the explanation.
I think we don't need to check `!addonHasFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer)` again, since we have checked this in [line 85](https://github.com/open-cluster-management-io/addon-framework/pull/226/files#diff-e4f38c04c06c31582696236d2246c4bf5d7543c79d1030ab466c42b466222117R85)

Reply:
Ah, that was a copy-paste error during development. Thanks for catching that!